### ISSUE TITLE
Add option for exhaustive union queries

### ISSUE THREAD

**Issue by amiralies** [reasonml-community/graphql-ppx#225]
Following discussions in discord I think it makes sense to have an option to do exhaustive union queries.

let's say I have a union type
```gql
union T = A | B | C
```
I want to get a warning/error when I try to query this type without covering all branches.

```gql
# C is missing
{
  queryT {
    ... on A {
      __typename
    }
    ... on B {
      __typename
    }
  }
}
```

How to enable this option?

We can have an option on bsconfig `{ "exhaustiveUnions": true }` and also a directive `@exahustive` for cases when people don't want to do exhaustive all union queries. we can also have a `@nonexhaustive` directive for suppressing warning/error on a single query when the project wide option is enabled.

**Comment by amiralies** [comment]
I'm interested in working on this, though I'm not very familiar with the codebase, @jfrolich is this possible for you to share guide /tips on what should be done?

**Comment by jfrolich** [comment]
Hey, yes that would be amazing. Unfortunately I don't have a lot of time with a newborn. It would be great to have some kind of guide at some point. Most of the difficulty in the codebase is dealing with the OCaml parse tree (ast). I recently started to create `rei` files to have a clearer public interface with types. I don't really know where to start explaining. I would suggest to dive in and try to change a tiny thing and go from there.

**Comment by jfrolich** [comment]
I would also have to dive in to see how the codebase is tracking exhaustiveness (has been some time ago).

**Comment by jfrolich** [comment]
Probably good to look at `generate_poly_variant_union_decoder` in `output_bucklescript_parser.re`

**Comment by amiralies** [comment]
Thank you, I will take a look

**Comment by amiralies** [comment]
Looking at native code, it seems like when querying a union non exhaustivly it generates another variant (`Nonexhaustive`) alongside varians associated with union type arms.
What do you think about adding this to rescript side instead of throwing warning/error.
For sake of backward compatibility and customization we can only enable this via an option.

**Comment by jfrolich** [comment]
Yeah I think it makes sense to have that instead of FutureAddedValue when it's not exhaustive.
